### DISCUSSION TITLE
How to detect device type or width on ssr?

### DISCUSSION THREAD

**Discussion by loremru** [nuxt/nuxt#17603] [Questions] [answered]
In my case i should detect window width to render necessary template. 
But in ssr there isnt Window, so I have two questions:

1) How can i get User-Agent and save device type to my store?
2) Why is there no way to provide some window options to SSR via Headers? (like innerWidth)

**Comment by KennethTrecy** [comment]
Try to use `process.client` inside setup.
```
<template>
  // Your tags here...
</template>

<script setup>
const innerW = ref(0)
if (process.client) {
	console.log(navigator.userAgent)
	console.log(window.innerWidth)
	setTimeout(() => {
		innerW.value = window.innerWidth
	}, 500)
}

setTimeout(() => {
	console.log("width is " + innerW.value)
}, 1000)
</script>
```

**Reply by loremru** [reply]
> Try to use `process.client` inside setup.
> 
> ```
> <template>
>   // Your tags here...
> </template>
> 
> <script setup>
> if (process.client) {
>   console.log(navigator.userAgent)
>   console.log(window.innerWidth)
> }
> </script>
> ```

In this case, the SSR is useless. The template is rendered on the server and on the server I need to know the type of device

**Reply by KennethTrecy** [reply]
You may also use `useRequestHeaders()`:
```
<template>
  // Your tags here...
</template>

<script setup>
const request = useRequestHeaders();

console.log(request['user-agent'])
</script>
```

**Comment by TheAlexLichter** [comment, marked as answer]
>     1. How can i get User-Agent and save device type to my store?

On the server side you can use the `useReqHeaders` composable, on the client side `navigator.userAgent`.
Check my [example here](https://codesandbox.io/s/hungry-glitter-ruf7lk?file=/app.vue)

Besides the UA, you can also check if the browser sends [client hints](https://developer.mozilla.org/en-US/docs/Web/HTTP/Client_hints).

>     2. Why is there no way to provide some window options to SSR? (like innerWidth via headers)
Because there is no `window` on the server. It is a browser-specific API that is only available on client side.

**Reply by loremru** [reply]
Thanks that exactly what i looking for.

But second question is still open. If there was a method to extend these headers with window.innerWidth, that would be great

**Reply by TheAlexLichter** [reply]
> But second question is still open. If there was a method to extend these headers with window.innerWidth, that would be great

As I mentioned, this is not possible because there is no `window` on the server-side and properties like `innerWidth` are not known at this point in time.

If the browser supports it, it may send [client hints](https://developer.mozilla.org/en-US/docs/Web/HTTP/Client_hints) (also linked above) that provide info about the client window, e.g. the `width`, via headers. But this is highly experimental.

Besides that, I'm not aware of any other options to find out more server-side as it is "too early".
